File: frontend/src/services/plans.ts
import api from "./api"

export interface Plan {
  id: string
  name: string
  description: string | null
  type: string
  download_speed_mbps: number
  upload_speed_mbps: number
  data_cap_gb: number | null
  price: number
  setup_fee: number
  billing_cycle: string
  is_taxable: boolean
  is_active: boolean
}

export interface PlanCreate {
  name: string
  type: string
  download_speed_mbps: number
  upload_speed_mbps: number
  price: number
  setup_fee?: number
  billing_cycle?: string
  data_cap_gb?: number
  description?: string
}

export async function listPlans(params?: { skip?: number; limit?: number }) {
  const res = await api.get("/plans", { params })
  return res.data as Plan[]
}

export async function getPlan(id: string) {
  const res = await api.get(`/plans/${id}`)
  return res.data as Plan
}

export async function createPlan(data: PlanCreate) {
  const res = await api.post("/plans", data)
  return res.data as Plan
}

export async function updatePlan(id: string, data: Partial<PlanCreate>) {
  const res = await api.put(`/plans/${id}`, data)
  return res.data as Plan
}

export async function deletePlan(id: string) {
  await api.delete(`/plans/${id}`)
}

File: backend/app/api/v1/plans.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import AdminOnly, AnyStaff
from app.models.user import User
from app.schemas.plan import PlanCreate, PlanRead, PlanUpdate
from app.services.plan_service import PlanService

router = APIRouter(prefix="/plans", tags=["plans"])


@router.get("", response_model=list[PlanRead])
async def list_plans(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AnyStaff),
):
    service = PlanService(db)
    return await service.list(user.organization_id, skip=skip, limit=limit)


@router.post("", response_model=PlanRead, status_code=201)
async def create_plan(
    data: PlanCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AdminOnly),
):
    service = PlanService(db)
    return await service.create(data, organization_id=user.organization_id)


@router.get("/{plan_id}", response_model=PlanRead)
async def get_plan(
    plan_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AnyStaff),
):
    service = PlanService(db)
    return await service.get(plan_id, user.organization_id)


@router.put("/{plan_id}", response_model=PlanRead)
async def update_plan(
    plan_id: uuid.UUID,
    data: PlanUpdate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AdminOnly),
):
    service = PlanService(db)
    return await service.update(plan_id, data, user.organization_id)


@router.delete("/{plan_id}", status_code=204)
async def delete_plan(
    plan_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(AdminOnly),
):
    service = PlanService(db)
    await service.delete(plan_id, user.organization_id)
